types/wlr_linux_dmabuf.c

```diff
@@ -0,0 +1,435 @@
+				buffer->attributes.offset[0]);
+			goto err_out;
+		}
+
+		if (buffer->attributes.offset[0] + buffer->attributes.stride[0] > size) {
+			wl_resource_post_error(params_resource,
+				ZWP_LINUX_BUFFER_PARAMS_V1_ERROR_OUT_OF_BOUNDS,
+				"invalid stride %i for plane",
+				buffer->attributes.stride[0]);
+			goto err_out;
+		}
+
+		if (buffer->attributes.offset[0] + buffer->attributes.stride[0] * height > size) {
+			wl_resource_post_error(params_resource,
+				ZWP_LINUX_BUFFER_PARAMS_V1_ERROR_OUT_OF_BOUNDS,
+				"invalid buffer stride or height for plane");
+				goto err_out;
+		}
+	}
+
+	/* reject unknown flags */
+	if (buffer->attributes.flags & ~ZWP_LINUX_BUFFER_PARAMS_V1_FLAGS_Y_INVERT) {
+		wl_resource_post_error(params_resource,
+				ZWP_LINUX_BUFFER_PARAMS_V1_ERROR_INVALID_FORMAT,
+				"Unknown dmabuf flags %"PRIu32, buffer->attributes.flags);
+		goto err_out;
+	}
+
+	/* Check if dmabuf is usable */
+	if (!wlr_egl_check_import_dmabuf(buffer->egl, buffer)) {
+		goto err_failed;
+	}
+
+	buffer->buffer_resource = wl_resource_create(client, &wl_buffer_interface,
+		1, buffer_id);
+	if (!buffer->buffer_resource) {
+		wl_resource_post_no_memory(params_resource);
+		goto err_failed;
+	}
+
+	wl_resource_set_implementation(buffer->buffer_resource,
+		&wl_buffer_impl, buffer, handle_buffer_destroy);
+
+	/* send 'created' event when the request is not for an immediate
+	 * import, that is buffer_id is zero */
+	if (buffer_id == 0) {
+		zwp_linux_buffer_params_v1_send_created(params_resource,
+			buffer->buffer_resource);
+	}
+	return;
+
+err_failed:
+	if (buffer_id == 0) {
+		zwp_linux_buffer_params_v1_send_failed(params_resource);
+	} else {
+		/* since the behavior is left implementation defined by the
+		 * protocol in case of create_immed failure due to an unknown cause,
+		 * we choose to treat it as a fatal error and immediately kill the
+		 * client instead of creating an invalid handle and waiting for it
+		 * to be used.
+		 */
+		wl_resource_post_error(params_resource,
+			ZWP_LINUX_BUFFER_PARAMS_V1_ERROR_INVALID_WL_BUFFER,
+			"importing the supplied dmabufs failed");
+	}
+err_out:
+	linux_dmabuf_buffer_destroy(buffer);
+	return;
+}
+
+static void params_create(struct wl_client *client,
+		struct wl_resource *params_resource,
+		int32_t width, int32_t height,uint32_t format, uint32_t flags) {
+	params_create_common(client, params_resource, 0, width, height, format, flags);
+}
+
+static void params_create_immed(struct wl_client *client,
+		struct wl_resource *params_resource, uint32_t buffer_id,
+		int32_t width, int32_t height,uint32_t format, uint32_t flags) {
+	params_create_common(client, params_resource, buffer_id, width, height, format, flags);
+}
+
+static const struct zwp_linux_buffer_params_v1_interface linux_buffer_params_impl = {
+	params_destroy,
+	params_add,
+	params_create,
+	params_create_immed,
+};
+
+struct wlr_dmabuf_buffer *wlr_dmabuf_buffer_from_params_resource(
+	struct wl_resource *params_resource) {
```

Comment:
Style nit: missing an extra tab here, so we see the difference between the function signature and the function body.